### PR TITLE
Auth: Signing Key persistence

### PR DESCRIPTION
**Pull request by Jguer**
**What is this feature?**

Allows instances to persist signing keys for onprem use

Missing from this PR:

- [x] Background key rotation
- [x] Remote Cache for JWKS

**Why do we need this feature?**

[Add a description of the problem the feature is trying to solve.]

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/services/signingkeys/signingkeysimpl/service.go [hunk 1]
```diff
@@ -30,7 +31,10 @@ func ProvideEmbeddedSigningKeysService() (*Service, error) {
 		return nil, signingkeys.ErrKeyGenerationFailed.Errorf("Error generating private key: %v", err)
 	}
 
-	if err := s.AddPrivateKey(serverPrivateKeyID, privateKey); err != nil {
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
+	defer cancel()
+	expiry := time.Now().Add(30 * 24 * time.Hour)
+	if err := s.AddPrivateKey(ctx, signingkeys.ServerPrivateKeyID, privateKey, jose.ES256, &expiry, false); err != nil {
```
Comment (Jguer) [inline pkg/services/signingkeys/signingkeysimpl/service.go:37]: @chagrafana @jmatosgrafana 

If you have views on the default rotation of signing keys (set to 30 days currently) and the algo, please comment here

File: pkg/services/sqlstore/migrations/signingkeys/migrations.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import "github.com/grafana/grafana/pkg/services/sqlstore/migrator"
+
+func AddMigration(mg *migrator.Migrator) {
+	var signingKeysV1 = migrator.Table{
+		Name: "signing_key",
+		Columns: []*migrator.Column{
+			{Name: "id", Type: migrator.DB_BigInt, IsPrimaryKey: true, IsAutoIncrement: true},
+			{Name: "key_id", Type: migrator.DB_NVarchar, Length: 255, Nullable: false},
+			{Name: "private_key", Type: migrator.DB_Text, Nullable: false},
+			{Name: "added_at", Type: migrator.DB_DateTime, Nullable: false},
```
Comment (eleijonmarck) [inline pkg/services/sqlstore/migrations/signingkeys/migrations.go:12]: nit: `added_at`, is not this the same as `created_at`.

to be more consistent with column names

Comment (Jguer) [reply pkg/services/sqlstore/migrations/signingkeys/migrations.go:12]: Is it understandable it's only the time the key was added to the DB and not it's actual creation time?

File: pkg/services/signingkeys/signingkeystore/store.go [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+				key.KeyID, key.PrivateKey, key.AddedAt, key.Alg, key.ExpiresAt)
+			return err
+		}
+
+		if force || (existingKey.ExpiresAt != nil && existingKey.ExpiresAt.Before(time.Now())) {
+			_, err = tx.Exec(ctx,
+				"UPDATE signing_key SET private_key = ?, added_at = ?, alg = ?, expires_at = ? WHERE key_id = ?",
+				key.PrivateKey, key.AddedAt, key.Alg, key.ExpiresAt, key.KeyID)
+			return err
+		}
```
Comment (kalleep) [inline pkg/services/signingkeys/signingkeystore/store.go:112]: I guess this is intended to rotate keys right?

When a key rotation happen we should probably keep the old key for a while so we don't immateriality invalidate everything signed with the previous key, WDYT?

**Review by kalleep — COMMENTED** [review]
Looks good, left a question I think we need to decide on.

File: pkg/services/auth/idimpl/signer.go [hunk 4]
```diff
@@ -13,11 +13,15 @@ import (
 var _ auth.IDSigner = (*LocalSigner)(nil)
 
 func ProvideLocalSigner(keyService signingkeys.Service) (*LocalSigner, error) {
-	key := keyService.GetServerPrivateKey() // FIXME: replace with signing specific key
+	key, err := keyService.GetPrivateKey(context.Background(), signingkeys.ServerPrivateKeyID)
```
Comment (mgyongyosi) [inline pkg/services/auth/idimpl/signer.go:16]: Just a note: I think (in the future) it would be safer to have a separate key for this signer.

Comment (kalleep) [reply pkg/services/auth/idimpl/signer.go:16]: yes It is going to. before this pr it was not possible. So will make a pr to fix it in the id signer after we get this one merged :)

File: pkg/services/signingkeys/signingkeystore/store.go [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+	decrypted, err := s.secretsService.Decrypt(ctx, payload)
+	if err != nil {
+		return nil, err
+	}
+
+	block, _ := pem.Decode(decrypted)
+	if block == nil {
+		return nil, errors.New("failed to decode private key PEM")
+	}
+
```
Comment (mgyongyosi) [inline pkg/services/signingkeys/signingkeystore/store.go:183]: Just to be more defensive I think a check for block type would make sense, wdyt?
```
if block.Type != "PRIVATE KEY" {
	return nil, errors.New("invalid block type")
}
```

File: pkg/services/signingkeys/signingkeystore/store.go (hunk 3 above)
Comment (mgyongyosi) [reply pkg/services/signingkeys/signingkeystore/store.go:112]: +1, maybe introducing a `version` field to the `signing_key` table? (And eventually the JWT header's `kid` should contain the version as well, something like `keyname-1`)

File: pkg/services/sqlstore/migrations/signingkeys/migrations.go (hunk 2 above)
Comment (eleijonmarck) [reply pkg/services/sqlstore/migrations/signingkeys/migrations.go:12]: is there a creation time then?

File: pkg/services/signingkeys/signingkeystore/store.go (hunk 3 above)
Comment (Jguer) [reply pkg/services/signingkeys/signingkeystore/store.go:112]: +1, that's a good point, we need to account for both keys existing at the same time. But I guess we'll never rotate them (no update), as they need to exist as independent entries. Let me see if I can find a way to make it work

File: pkg/services/signingkeys/signingkeystore/store.go (hunk 5 above)
Comment (Jguer) [reply pkg/services/signingkeys/signingkeystore/store.go:183]: +1

File: pkg/services/signingkeys/signingkeystore/store.go [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+		return nil, err
+	}
+
+	// Encode private key to PEM format
+	privateKeyPEM := pem.EncodeToMemory(&pem.Block{
+		Type:  "PRIVATE KEY",
+		Bytes: pKeyBytes,
+	})
+
+	encrypted, err := s.secretsService.Encrypt(ctx, privateKeyPEM, secrets.WithoutScope())
```
Comment (jmatosgrafana) [inline pkg/services/signingkeys/signingkeystore/store.go:152]: which encryption key is behind `secrets.WithoutScope()` ?

File: pkg/services/signingkeys/signingkeysimpl/service.go (hunk 1 above)
Comment (jmatosgrafana) [reply pkg/services/signingkeys/signingkeysimpl/service.go:37]: Algortithm is fine (ECDSA 256 bit) and 30 days retention is OK as long as we trust the encrypted storage in database.
I left a comment related to that [here](https://github.com/grafana/grafana/pull/75487/files#r1338660797).
It's all about the key management for the encryption of this private key

File: pkg/services/signingkeys/signingkeystore/store.go (hunk 6 above)
Comment (Jguer) [reply pkg/services/signingkeys/signingkeystore/store.go:152]: the `root` key, same one used by oauth token encryption and secret kvstore. 

It's either generated on the fly or uses a different secrets backend from my fast research

https://github.com/grafana/grafana/blob/main/pkg/services/secrets/manager/manager.go#L187

File: pkg/services/signingkeys/signingkeys.go [hunk 7]
```diff
@@ -8,27 +8,29 @@
...
-	// GetServerPrivateKey returns the private key used to sign tokens
-	GetServerPrivateKey() crypto.PrivateKey
-	// GetServerPublicKey returns the public key used to verify tokens
-	GetServerPublicKey() crypto.PublicKey
+	GetPrivateKey(ctx context.Context, keyID string) (crypto.PrivateKey, error)
+	// GetPublicKey returns the public key with the specified key ID
+	GetPublicKey(ctx context.Context, keyID string) (crypto.PublicKey, error)
 	// AddPrivateKey adds a private key to the service
-	AddPrivateKey(keyID string, privateKey crypto.PrivateKey) error
+	AddPrivateKey(ctx context.Context, keyID string,
```
Comment (kalleep) [inline pkg/services/signingkeys/signingkeys.go:33]: Make these private, we don't need them externally right now

**Review by kalleep — APPROVED** [review]
Lets get this merge and address the things we discussed in call in smaller pr:s :)

**Pull request merged**